xonsh/base_shell.py

```diff
@@ -373,13 +373,21 @@ def _append_history(self, tee_out=None, **info):
 
     def _fix_cwd(self):
         """Check if the cwd changed out from under us"""
+        env = builtins.__xonsh_env__
         cwd = os.getcwd()
-        if cwd != builtins.__xonsh_env__.get('PWD'):
-            old = builtins.__xonsh_env__.get('PWD')  # working directory changed without updating $PWD
-            builtins.__xonsh_env__['PWD'] = cwd      # track it now
+        if cwd != env['PWD']:
```

Comment:
You need to check that PWD exists or use something like `env.get('PWD', None)`. `$PWD` is not in the list of variables that the `Env` class handles. Maybe it should be, but that can be a different PR. 

The same comment applies to all the other places you access `$PWD` below.